### PR TITLE
Refactor subnet lease

### PR DESCRIPTION
**Pull request by gbenhaim**
The general idea of how to manage the store has not
changed, but the following changes were made:

1. Created a 'SubnetStore' class which represents the store.
   (it makes unit testing easier)

2. Created a 'Lease' class which represents a lease file in the store.

3. Allow requesting a specific lease from the store.

4. Increased the range of the store to 192.168.200.0-192.168.255.0

5. Use 'netaddr' to validate and represent ip addresses.

6. Improved exception handling.

7. Added unit tests.

closes https://github.com/lago-project/lago/issues/530
closes https://github.com/lago-project/lago/issues/523
closes https://github.com/lago-project/lago/issues/454

Signed-off-by: gbenhaim <galbh2@gmail.com>

### CODE REVIEW

**Comment by gbenhaim** [comment]
ci test please

**Comment by gbenhaim** [comment]
ci test please

**Comment by gbenhaim** [comment]
ci test please

File: lago/subnet_lease.py [hunk 1]
```diff
@@ -17,222 +17,505 @@
...
-
-def _lease_valid(path):
+    SubnetStore object represents a store of subnets used by lago for network
+    bridges.
+
+    .. note:: Currently only /24 ranges are handled, and all of them under the
+        192.168._min_third_octet to 192.168._max_third_octet ranges.
+
+
+    The leases are stored under :class:`LEASE_DIR` as json files with the
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: That is a class? LEASE_DIR?

File: lago/subnet_lease.py [hunk 2]
```diff
@@ -17,222 +17,505 @@
...
+
+        [
+            "/path/to/prefix/uuid/file",
+            "uuid_hash",
+        ]
+
+    Where the `uuid_hash` is the 32 char uuid of the prefix (the contents of
+    the uuid file at the time of doing the lease).
+
+    The helper class :class:'Lease' is used to abstract the interaction with
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: I think you need to change ' to -> ` for it to format in the docs.

File: lago/subnet_lease.py [hunk 3]
```diff
@@ -17,222 +17,505 @@
...
+            self._subnet_template.format(min_third_octet)
+        )
+        self._max_subnet = IPNetwork(
+            self._subnet_template.format(max_third_octet)
+        )
+        self._validate_lease_dir()
+
+    def _create_lock(self):
+        return utils.LockFile(
+            path=os.path.join(self.path, LOCK_NAME), timeout=5
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: cool, the time out promises us retries?

File: lago/subnet_lease.py [hunk 4]
```diff
@@ -17,222 +17,505 @@
...
+
+        Raises:
+            LagoSubnetLeaseException:
+                1. If this store is full
+                2. If the requested subnet is already taken.
+            LagoSubnetLeaseLockException:
+                If the lock to self.path can't be acquired.
+        """
+        try:
+            with self._create_lock():
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: I think this can be improved a little(but not critical for this PR) - putting the lock exactly where the critical code is - when writing the subnet lease, and when reading it. Also, I'm not 100% sure you need a lock for the reading part.

File: lago/subnet_lease.py [hunk 5]
```diff
@@ -17,222 +17,505 @@
...
+
+        with open(uuid_path, mode='rt') as f:
+            if f.read() == uuid:
+                return uuid_path
+            else:
+                return False
+
+    @property
+    def exist(self):
+        return os.path.isfile(self.path)
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: This will throw a nasty exception if the user does not have read access to that path.
In the longer run we should discuss how can we re-factor it to support multi-users..

For now I think adding a try/except here and assuming that if you don't have read access it just exists would be fine, no?

File: lago/subnet_lease.py [hunk 6]
```diff
@@ -17,222 +17,505 @@
...
+        Returns:
+            bool: True if subnet can be parsed into IPNetwork object and is
+                inside the range, False otherwise
+        """
+        try:
+            return \
+                self._min_subnet <= \
+                IPNetwork(subnet) <= \
+                self._max_subnet
+        except AddrFormatError:
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: Doesn't this mask the error if the subnet format was given improperly?

File: lago/subnet_lease.py [hunk 7]
```diff
@@ -17,222 +17,505 @@
...
+class LagoSubnetLeaseLockException(LagoSubnetLeaseException):
+    def __init__(self, store_path):
+        super(LagoSubnetLeaseLockException, self).__init__(
+            dedent(
+                """
+                Failed to acquire a lock for store {}.
+                This failure can be caused by several reasons:
+                1. Another 'lago' environment is using the store.
+                2. A stale lock was left in the store.
+                3. The group which owns the path to the store isn't 'lago',
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: Maybe just say the user doesn't have access to the store? because if he installed from pip - he didn't necessary setup a lago group at all(and also we don't really know if that is the reason - just that we can't access it).

File: lago/subnet_lease.py [hunk 8]
```diff
@@ -17,222 +17,505 @@
...
+
+    def __str__(self):
+        return self.subnet
+
+
+class LagoSubnetLeaseException(utils.LagoException):
+    pass
+
+
+class LagoSubnetLeaseLockException(LagoSubnetLeaseException):
```
Comment (nvgoldin) [inline lago/subnet_lease.py]: Cool! the exceptions messages really make it user friendly.

File: tests/unit/lago/test_subnet_lease.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            len(filenames) == 1 and \
+            lease_has_valid_uuid(path_to_lease)
+
+    @pytest.mark.parametrize('subnet', ['256.256.256.56', '10.10.10.0'])
+    def test_fail_on_out_of_range_subnet(self, subnet_store, subnet, prefix):
+        with pytest.raises(subnet_lease.LagoSubnetLeaseOutOfRangeException):
+            subnet_store.acquire(prefix.uuid_path, subnet)
+
+    def test_fail_on_full_store(self, subnet_store, prefix):
+        with pytest.raises(subnet_lease.LagoSubnetLeaseStoreFullException):
```
Comment (nvgoldin) [inline tests/unit/lago/test_subnet_lease.py]: This could fail earlier, I would split it:
1. First loop and fill the store
2. Then use pytest.raises on the the final operation you expect the failure

File: tests/unit/lago/test_subnet_lease.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+            for i in xrange(
+                subnet_store._min_third_octet,
+                subnet_store._max_third_octet + 2
+            ):
+                subnet_store.acquire(prefix.uuid_path)
+
+    def test_recalim_lease(self, subnet_store, prefix):
+        network = subnet_store.acquire(prefix.uuid_path)
+        recalimed_network = subnet_store.acquire(
```
Comment (nvgoldin) [inline tests/unit/lago/test_subnet_lease.py]: s/recalimed/reclaimed

File: lago.spec.in [hunk 11]
```diff
@@ -79,6 +79,7 @@ BuildRequires: python-dulwich
 %endif
 BuildRequires: python-flake8
 BuildRequires: git
+BuildRequires: python-netaddr
```
Comment (gbenhaim) [inline lago.spec.in]: Because of the dynamic generation of the config:
```%{buildroot}/usr/bin/lago --ignore-warnings generate-config > "%{buildroot}/%{_sysconfdir}/lago/lago.conf"```

File: lago/subnet_lease.py (hunk 1 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: Fixed.

File: lago/subnet_lease.py (hunk 2 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: Fixed.

File: lago/subnet_lease.py (hunk 3 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: LockFile will take care of retries in intervals of 0.1 second (this value is hardcoded).

File: lago/subnet_lease.py (hunk 4 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: The interface of the store doesn't provide a method for only reading its content, and even if it was, a lock should be created while reading, otherwise, the store can be changed while reading.

Comment (nvgoldin) [inline lago/subnet_lease.py]: Right - but you could:
1. Find your free subnet - this requires only reading(without a lock)
2. Get the *write* lock and attempt to acquire the subnet, if you failed, go back to 1.
Which is probably more efficient as more threads(i.e. lago init commands) would try to lock.

File: lago.spec.in (hunk 11 above)
Comment (nvgoldin) [inline lago.spec.in]: Right :)
always forget about that..

File: lago/subnet_lease.py (hunk 6 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: Fixed. 
Also separated the test to two, one for malformed address and the other for out of range address.

File: lago/subnet_lease.py (hunk 5 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: Done

File: lago/subnet_lease.py (hunk 8 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: Thanks :)

File: lago/subnet_lease.py (hunk 7 above)
Comment (gbenhaim) [inline lago/subnet_lease.py]: Done

File: tests/unit/lago/test_subnet_lease.py (hunk 9 above)
Comment (gbenhaim) [inline tests/unit/lago/test_subnet_lease.py]: Done

File: tests/unit/lago/test_subnet_lease.py (hunk 10 above)
Comment (gbenhaim) [inline tests/unit/lago/test_subnet_lease.py]: Done

**Comment by gbenhaim** [comment]
ci merge please
